From 2c3a1fbb7683569b16b860aa22322cf4c490b650 Mon Sep 17 00:00:00 2001
From: Sathishkumar Muruganandam <murugana@qti.qualcomm.com>
Date: Tue, 21 Apr 2026 17:16:51 +0530
Subject: [PATCH] UPSTREAM:PROTOCOL: skip idle STA interfaces in duplicate link
 address check

When checking for duplicate link addresses, STA interfaces that are idle
with wpa_supplicant running but not associated yet,should be ignored.

Add a condition to skip `other_sdata` if it is STA sdata running but not
associated yet.

Currently all -ENOTUNIQ rejection don't have error prints to show interface
or link information whose addresses are duplicate. Added sdata_err prints
for the same.

Fixes: I7ba666144 (UPSTREAM:PROTOCOL: Reject duplicate per-link MACs
across interfaces)

Patch-dependency: I7ba666144 (UPSTREAM:PROTOCOL: Reject duplicate per-link MACs
across interfaces)

Change-Id: I5ca696c79c6c1249ce0a09fbecaa63666003e74b
Signed-off-by: Sathishkumar Muruganandam <murugana@qti.qualcomm.com>
---
 net/mac80211/iface.c |  6 +++++-
 net/mac80211/link.c  | 22 ++++++++++++++++++--
 2 files changed, 25 insertions(+), 3 deletions(-)

diff --git a/net/mac80211/iface.c b/net/mac80211/iface.c
index 6da5550942b4..84d005da715a 100644
--- a/net/mac80211/iface.c
+++ b/net/mac80211/iface.c
@@ -377,8 +377,12 @@ static int ieee80211_check_concurrent_iface(struct ieee80211_sub_if_data *sdata,
			 * check whether it may have the same address
			 */
			if (!identical_mac_addr_allowed(iftype,
-							nsdata->vif.type))
+							nsdata->vif.type)) {
+				sdata_err(sdata, "DUP MAC addr %pM -> %s,vif types: %d/%d\n",
+                                          sdata->vif.addr, nsdata->name,
+                                          sdata->vif.type, nsdata->vif.type);
				return -ENOTUNIQ;
+			}

			/*
			 * can only add VLANs to enabled APs
diff --git a/net/mac80211/link.c b/net/mac80211/link.c
index 64970e4be20b..efe8a47630b2 100644
--- a/net/mac80211/link.c
+++ b/net/mac80211/link.c
@@ -278,13 +278,27 @@ static int ieee80211_check_dup_link_addrs(struct ieee80211_sub_if_data *sdata)
			if (is_zero_ether_addr(addr1))
				continue;

+			/* Skip idle STAs (wpa_supplicant running but not associated) */
+			if (other_sdata->vif.type == NL80211_IFTYPE_STATION) {
+				struct ieee80211_if_managed *other_ifmgd =
+					&other_sdata->u.mgd;
+
+				if (!other_ifmgd->associated)
+					continue;
+			}
+
			/* Compare link address against other_sdata's MLD address */
			addr2 = other_sdata->vif.addr;
			if (!is_same_sdata && !is_zero_ether_addr(addr2) &&
			    ether_addr_equal(addr1, addr2) &&
			    !identical_mac_addr_allowed(sdata->vif.type,
-							other_sdata->vif.type))
+							other_sdata->vif.type)) {
+				sdata_err(sdata, "DUP MAC link[%d] addr %pM -> %s,valid_links=0x%x vif types: %d/%d\n",
+					  link_id, addr2, other_sdata->name,
+					  other_sdata->vif.valid_links,
+					  sdata->vif.type, other_sdata->vif.type);
				return -ENOTUNIQ;
+			}

			for (o_link_id = 0;
			     o_link_id < IEEE80211_MLD_MAX_NUM_LINKS;
@@ -309,8 +323,12 @@ static int ieee80211_check_dup_link_addrs(struct ieee80211_sub_if_data *sdata)
				if (ether_addr_equal(addr1, addr2) &&
				    (is_same_sdata ||
				    !identical_mac_addr_allowed(sdata->vif.type,
-								other_sdata->vif.type)))
+								other_sdata->vif.type))) {
+					sdata_err(sdata, "DUP MAC link[%d] addr %pM -> %s:link[%d], vif types: %d/%d\n",
+						  link_id, addr2, other_sdata->name, o_link_id,
+						  sdata->vif.type, other_sdata->vif.type);
					return -ENOTUNIQ;
+				}
			}
		}
	}
--
2.34.1
